Add Cluster Health metrics

h3llix opened the pull request:
See commit logs for details.(a screen shot has also been attached
![final](https://user-images.githubusercontent.com/55123900/111907663-587b8580-8a7c-11eb-974d-05095a39fd8f.png)

Fixes : #15269

h3llix committed:
docs: add docs for cluster health

Docs for unreachable_nodes  and unreachable_health_endpoints have been added.

Fixes : #15269

Signed-off-by: Gaurav Genani <h3llix.pvt@gmail.com>

joestringer commented on Documentation/operations/metrics.rst:
```diff
@@ -196,6 +196,7 @@ Name                                         Labels
 ``endpoint_regenerations_total``             ``outcome``                                        Count of all endpoint regenerations that have completed
 ``endpoint_regeneration_time_stats_seconds`` ``scope``                                          Endpoint regeneration time stats
 ``endpoint_state``                           ``state``                                          Count of all endpoints
+`unreachable_health_endpoints``                                                                 Number of health endpoints that cannot be reached
```
Looks like there's a ` character missing at the start here.

h3llix replied:
@joestringer @pchaigno thanks for the feedback . And yes this looks appealing that they should lie under a new category . I will fix this and update the pr

joestringer reviewed (changes requested):
In addition to the feedback from Paul (+ below nit), I think that both of these metrics should be in a new category called something like "cluster health" rather than "endpoints" and "nodes". For the target audience of these docs, it's probably more interesting to see the heading "Cluster health" and figure out which metrics to monitor / alert to determine that cluster health is good rather than trying to piece together the intent from separate categories.

pchaigno commented on Documentation/operations/metrics.rst:
```diff
@@ -207,6 +207,16 @@ Name                                       Labels
 ``services_events_total``                                                                     Number of services events labeled by action type
 ========================================== ================================================== ========================================================
 
+Cilium health
```
I think this should be:
```suggestion
Cluster health
```

h3llix committed:
doc :  add metrics for cluster health

Eariler documentation for metrics `controllers_failing` and `ip_address` were missing

Signed-off-by: Gaurav Genani <h3llix.pvt@gmail.com>

pchaigno commented on Documentation/operations/metrics.rst:
```diff
@@ -207,6 +207,18 @@ Name                                       Labels
+========================================== ================================================== ========================================================
+``unreachable_nodes``                                                                         Number of nodes that cannot be reached
+``unreachable_health_endpoints``                                                              Number of health endpoints that cannot be reached
+``ip_addresses``                                                                              Number of allocated IP addresses
```
That metric seems to have a label.

h3llix replied:
> @pchaigno @joestringer Can you please have a look and tell me if the metrics are listed under correct headings , and how the labels are assigned to corresponding metrices ?

Sorry but i asked this earlier also . How to find the labels of the metrics ?

pchaigno replied:
```
$ git grep "func NewDesc(" vendor/github.com/prometheus/
vendor/github.com/prometheus/client_golang/prometheus/desc.go:func NewDesc(fqName, help string, variableLabels []string, constLabels Labels) *Desc {
```
I'm not sure if `variableLabels` and `constLabels` both need to be documented but checking existing entries should answer that question. Then:
```
$ git grep -C3 ip_addresses pkg/metrics/
pkg/metrics/status.go-                  nil, nil,
pkg/metrics/status.go-          ),
pkg/metrics/status.go-          ipAddressesDesc: prometheus.NewDesc(
pkg/metrics/status.go:                  prometheus.BuildFQName(Namespace, "", "ip_addresses"),
pkg/metrics/status.go-                  "Number of allocated IP addresses",
pkg/metrics/status.go-                  []string{"family"}, nil,
pkg/metrics/status.go-          ),
$ git grep -C3 unreachable_nodes pkg/metrics/
pkg/metrics/status.go-                  []string{"family"}, nil,
pkg/metrics/status.go-          ),
pkg/metrics/status.go-          unreachableNodesDesc: prometheus.NewDesc(
pkg/metrics/status.go:                  prometheus.BuildFQName(Namespace, "", "unreachable_nodes"),
pkg/metrics/status.go-                  "Number of nodes that cannot be reached",
pkg/metrics/status.go-                  nil, nil,
pkg/metrics/status.go-          ),
$ git grep -C3 unreachable_health_endpoints pkg/metrics/
pkg/metrics/status.go-                  nil, nil,
pkg/metrics/status.go-          ),
pkg/metrics/status.go-          unreachableHealthEndpointsDesc: prometheus.NewDesc(
pkg/metrics/status.go:                  prometheus.BuildFQName(Namespace, "", "unreachable_health_endpoints"),
pkg/metrics/status.go-                  "Number of health endpoints that cannot be reached",
pkg/metrics/status.go-                  nil, nil,
pkg/metrics/status.go-          ),
$ git grep -C3 controllers_failing pkg/metrics/
pkg/metrics/status.go-          ciliumClient: ciliumClient,
pkg/metrics/status.go-          healthClient: healthClient,
pkg/metrics/status.go-          controllersFailingDesc: prometheus.NewDesc(
pkg/metrics/status.go:                  prometheus.BuildFQName(Namespace, "", "controllers_failing"),
pkg/metrics/status.go-                  "Number of failing controllers",
pkg/metrics/status.go-                  nil, nil,
pkg/metrics/status.go-          ),
```

h3llix committed:
docs: add labels to Cluster health metrics

Signed-off-by: Gaurav Genani <h3llix.pvt@gmail.com>

pchaigno commented on Documentation/operations/metrics.rst:
```diff
@@ -207,6 +207,19 @@ Name                                       Labels
+========================================== ================================================== ========================================================
+``unreachable_nodes``                                                                         Number of nodes that cannot be reached
+``unreachable_health_endpoints``                                                              Number of health endpoints that cannot be reached
+``ip_addresses``                           ``family``                                         Number of allocated IP addresses
```
Maybe the IPAM section would be a better place for this metric? /cc @joestringer

joestringer replied:
Yes, please put `ip_addresses` under IPAM.

qmonnet commented on Documentation/operations/metrics.rst:
```diff
@@ -207,6 +207,19 @@ Name                                       Labels
+``unreachable_health_endpoints``                                                              Number of health endpoints that cannot be reached
+``ip_addresses``                           ``family``                                         Number of allocated IP addresses
+``controllers_failing``                                                                       Number of failing controllers
+
```
```suggestion
```
No need for this empty line in the table.
Looks good from my side otherwise.

qmonnet reviewed (approved):
Looks good, thank you!
